### PR TITLE
SQL: selecting a literal from grouped by query generates error

### LINKED ISSUE #41951: SQL: COUNT() makes a function resolution in WHERE clause to fail
A single `COUNT(*)` in the select portion prevents the field attribute resolution in WHERE clause
```
DELETE test
PUT test/_doc/1
{
  "int": 10
}

POST _sql?format=txt
{
  "query": "SELECT COUNT(*) FROM test WHERE ABS(int) > 0 GROUP BY int"
}
```

The query above fails wtth the following error:

```
{
  "error": {
    "root_cause": [
      {
        "type": "verification_exception",
        "reason": "Found 1 problem(s)\nline 1:55: Unknown column [int], did you mean [int]?"
      }
    ],
    "type": "verification_exception",
    "reason": "Found 1 problem(s)\nline 1:55: Unknown column [int], did you mean [int]?"
  },
  "status": 400
}
```
However if we add something besides `COUNT(*)` to the select clause, it works:

```
POST _sql?format=txt
{
  "query": "SELECT COUNT(*), MAX(int) FROM test WHERE ABS(int) > 0 GROUP BY int"
}
```
Removing the function for the WHERE clause also works:

```
POST _sql?format=txt
{
  "query": "SELECT COUNT(*) FROM test WHERE int > 0 GROUP BY int"
}
```

This issue might be related to #41413 but the use case seems to be quite different since it requires calling a function in WHERE and doesn't use a literal.

### PR DESCRIPTION
**Pull request by codebird**
related to #41951 and #41413

### CODE REVIEW

**Comment by astefan** [comment]
@codebird thank you for raising the PR. We'll be looking at it, but there may be some time until then. There are several issues in flight related to literals in different scenarios and definitely this one is similar.

**Comment by codebird** [comment]
updates, now it resolves both #41413 #41951, but this has issues #34583

**Comment by codebird** [comment]
@astefan now all 3 issues are solved. 35483 is a bit hacky solution. I'll keep on looking into it... But well if you guys check it before I find anything...

**Comment by codebird** [comment]
@astefan  is this pull request dead? Shall I close it?

**Comment by matriv** [comment]
Could you please add some unit tests (`QueryTranslatorTests`) and some integration tests (agg.sql-spec) ?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 1]
```diff
@@ -455,6 +455,12 @@ static EsQueryExec fold(AggregateExec a, EsQueryExec exec) {
                 // literal
                 if (target.foldable()) {
                     queryC = queryC.addColumn(ne.toAttribute());
+                    if(a.groupings().isEmpty() == false){
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: This if is redundant since below you iterate over the groupings, so if empty => no iteration

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 2]
```diff
@@ -341,6 +343,17 @@ else if (plan instanceof Aggregate) {
                     List<Expression> groupings = a.groupings();
                     List<Expression> newGroupings = new ArrayList<>();
                     AttributeMap<Expression> resolved = Expressions.aliases(a.aggregates());
+
+                    //if all we're selecting are functions or literals, even if we're giving them aliases.
+                    if(a.aggregates().stream().filter(s -> s.children().size()>0
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Please use `.forEachUp()`, you can find an example here: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java#L638

Also please check your formatting regarding whitespaces around `(` and `)`, and around operators, e.g.: `size()>0` => `size() > 0`.

**Review by matriv — COMMENTED** [review]
Left a couple of comments

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java (hunk 2 above)
Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: not sure of the forEchUp usage.

**Comment by codebird** [comment]
Don't know what I was thinking to write integrity at the place of integration in my commit message

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 3]
```diff
@@ -455,6 +455,13 @@ static EsQueryExec fold(AggregateExec a, EsQueryExec exec) {
                 // literal
                 if (target.foldable()) {
                     queryC = queryC.addColumn(ne.toAttribute());
+                    //If we're only selecting literals, we have to fix the row count
+                    if(a.aggregates().stream().allMatch(s -> s.children().get(0) instanceof Literal)) {
+                        for (Expression grouping : a.groupings()) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: I don't think that's the correct place to add the groupings.
We are iterating over all `a.aggregates()`.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 4]
```diff
@@ -130,6 +133,29 @@ public void testTermEqualityAnalyzer() {
...
+        List<Expression> groupings = ((Aggregate) p).groupings();
+        assertTrue(groupings.get(0).resolved());
+        var agg = ((Aggregate) p).aggregates();
+        assertEquals((agg.get(0)).name(), "c");
+        assertEquals(((Count) ((Alias) agg.get(0)).child()).functionName(), "COUNT");
+    }
+    public void testLiteralWithGroupBy(){
+        LogicalPlan p = plan("SELECT 1 as t FROM test GROUP BY int");
+        assertTrue(p instanceof Aggregate);
+        List<Expression> groupings = ((Aggregate) p).groupings();
```
Comment (matriv) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: Can you please assert all `groupings` and `aggregates` and first their `size()`?
Please also test with more than 1 literal in the SELECT clause.

File: x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec [hunk 5]
```diff
@@ -563,5 +566,7 @@ SELECT MIN(salary) min, MAX(salary) max, gender g, languages l, COUNT(*) c FROM
 // group by with literal
 implicitGroupByWithLiteral
 SELECT 10, MAX("salary") FROM test_emp;
+literalWithGroupBy
+SELECT 1 FROM test_emp GROUP BY gender;
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec]: Here too, could you please add one more integ test with more than 1 literal and also scalars involved?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java (hunk 3 above)
Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: yep, 100% true.

**Comment by codebird** [comment]
@matriv pls review.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 6]
```diff
@@ -2112,7 +2113,7 @@ protected LogicalPlan rule(LogicalPlan plan) {
 
             plan.forEachDown(a -> {
                 List<Object> values = extractConstants(a.aggregates());
-                if (values.size() == a.aggregates().size() && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
+                if (values.size() == a.aggregates().size() && a.groupings().isEmpty() && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]:
```suggestion
                if (values.size() == a.aggregates().size() && a.groupings().isEmpty() == true && isNotQueryWithFromClauseAndFilterFoldedToFalse(a)) {
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 7]
```diff
@@ -586,7 +588,13 @@ else if (target.foldable()) {
                     }
                 }
             }
-
+            //If we're only selecting literals, we have to fix the row count
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]:
```suggestion
            // If we're only selecting literals, we have to fix the row count
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 8]
```diff
@@ -586,7 +588,13 @@ else if (target.foldable()) {
...
+            if(a.aggregates().stream().allMatch(s -> s.children().size() > 0 && s.children().get(0) instanceof Literal)) {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]:
```suggestion
            if (a.aggregates().stream().allMatch(e -> e.anyMatch(Expression::foldable))) {
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 9]
```diff
@@ -455,6 +456,7 @@ static EsQueryExec fold(AggregateExec a, EsQueryExec exec) {
                 // literal
                 if (target.foldable()) {
                     queryC = queryC.addColumn(ne.toAttribute());
+                    literalId = id;
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Maybe it's better to construct a `literalId` which concatenates all the ids.

File: x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec [hunk 10]
```diff
@@ -563,6 +567,8 @@ SELECT MIN(salary) min, MAX(salary) max, gender g, languages l, COUNT(*) c FROM
 // group by with literal
 implicitGroupByWithLiteral
 SELECT 10, MAX("salary") FROM test_emp;
+literalWithGroupBy
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec]: Please have the following tests:
- `literalWithGroupBy` (withoutAlias)
- `literalsWithGroupBy` (withoutAlias)
- `aliasedLiteralWithGroupBy` (can have mixed aliased and unaliased)
- `aliasedLiteralsWtihGroupBy` (can have mixed aliased and unaliased)
-  `literalsWithMultipleGroupBy` (aliased and unaliased literals, with more than one column in the `GROUP BY` which can also be a mix of aliased/unaliased)

File: x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec [hunk 11]
```diff
@@ -144,6 +144,10 @@ countDistinctAlias
 SELECT COUNT(DISTINCT hire_date) AS count FROM test_emp;
 countDistinctAndCountSimpleWithAlias
 SELECT COUNT(*) cnt, COUNT(DISTINCT first_name) as names, gender FROM test_emp GROUP BY gender ORDER BY gender;
+countWithFunctionFilterAndGroupBy
+SELECT COUNT(*) as c FROM test_emp WHERE ABS(salary) > 0 GROUP BY gender ORDER BY gender;
+countLiteralAndAliasWithFunctionFilterAndGroupBy
+SELECT 1 as t, gender as g, COUNT(*) as c FROM test_emp WHERE ABS(salary) > 0 GROUP BY g ORDER BY gender;
```
Comment (matriv) [inline x-pack/plugin/sql/qa/src/main/resources/agg.sql-spec]: Same here, please check more cases:

- aliased and unaliased literals
- more columns in group by (aliased and not aliased)
- more columns in the where

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 12]
```diff
@@ -341,6 +343,22 @@ else if (plan instanceof Aggregate) {
                     List<Expression> groupings = a.groupings();
                     List<Expression> newGroupings = new ArrayList<>();
                     AttributeMap<Expression> resolved = Expressions.aliases(a.aggregates());
+
+                    //if all we're selecting are functions or literals, even if we're giving them aliases.
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]:
```suggestion
                    // if all we're selecting are functions or literals, even if we're giving them aliases.
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 13]
```diff
@@ -341,6 +343,22 @@ else if (plan instanceof Aggregate) {
...
+                    final boolean[] updateResolved = {true};
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]:
```suggestion
                    final Holder<Boolean> updateResolved = new Holder(Boolean.TRUE);
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 14]
```diff
@@ -341,6 +343,22 @@ else if (plan instanceof Aggregate) {
...
+                    a.forEachUp(p -> p.forEachExpressionsUp(e -> {
+                        if (e instanceof Alias
+                            && ((Alias) e).child() instanceof Function == false
+                            && ((Alias) e).child() instanceof Literal == false) {
+                            updateResolved[0] = false;
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]:
```suggestion
                            updateResolved.set(Boolean.FALSE);
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 15]
```diff
@@ -341,6 +343,22 @@ else if (plan instanceof Aggregate) {
...
+                        }
+                    }));
+                    if(updateResolved[0]){
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]:
```suggestion
                    if (updateResolved.get() == true) {
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 16]
```diff
@@ -341,6 +343,22 @@ else if (plan instanceof Aggregate) {
                     List<Expression> groupings = a.groupings();
                     List<Expression> newGroupings = new ArrayList<>();
                     AttributeMap<Expression> resolved = Expressions.aliases(a.aggregates());
+
+                    //if all we're selecting are functions or literals, even if we're giving them aliases.
+                    final boolean[] updateResolved = {true};
+                    a.forEachUp(p -> p.forEachExpressionsUp(e -> {
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: I think here you want to check only the `a.aggregates()` and not all the expressions (including the grouping).
So probably you need the following :
```
a.aggregates().forEach(e -> e.forEachUp(f -> {
    if (f instanceof Function == false && f.foldable() == false) {
        updateResolved.set(Boolean.FALSE);
    }
}));
```

**Review by matriv — COMMENTED** [review]
Left some more comments.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java (hunk 9 above)
Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: why? we only need one of them no matter which

**Comment by matriv** [comment]
@elasticmachine ok to test

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java (hunk 9 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Sorry, I was wrong, It's fine as is, I just think that you can use directly `id`, no need of a new variable.

Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: id is set and used inside the for loop, and it is changing with the for loop, that's why I needed literalId

Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: I can move this `String id` outside the for loop and get rid of literalId, that's true. I'll check the failures, and do this at the same time.

**Comment by codebird** [comment]
@matriv I think bwc, and default-distro aren't from me. but ci/2 is Checkstyle failure, does that mean some issue with spacing or something like that?

**Comment by matriv** [comment]
@codebird You can run `./gradlew -p x-pack/plugin/sql checkstyleMain checkstyleTest`
Or to run everything (compile, checkstyle, unit tests, integ tests:
`./gradlew -p x-pack/plugin/sql check`

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 17]
```diff
@@ -341,6 +343,19 @@ else if (plan instanceof Aggregate) {
...
                     List<Expression> newGroupings = new ArrayList<>();
                     AttributeMap<Expression> resolved = Expressions.aliases(a.aggregates());
+
+                    final Holder<Boolean> updateResolved = new Holder<Boolean>(Boolean.FALSE);
+                    a.aggregates().forEach(e -> e.forEachUp(f -> {
+                        if (f instanceof Function == false && f.foldable() == false) {
+                            updateResolved.set(Boolean.TRUE);
+                        }
+                    }));
+                    if (updateResolved.get() == true){
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]:
```suggestion
                    if (updateResolved.get() == true) {
```

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java [hunk 18]
```diff
@@ -586,7 +587,13 @@ else if (target.foldable()) {
                     }
                 }
             }
-
+            // If we're only selecting literals, we have to fix the row count
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryFolder.java]: Maybe rephrase to something like`.. literals, we have to still execute the aggregation to create the correct grouping buckets, in order to return the appropriate number of rows`

**Comment by codebird** [comment]
@matriv what do I do with this: "Do not check for equality with true, since it is implied"?

**Comment by matriv** [comment]
> @matriv what do I do with this: "Do not check for equality with true, since it is implied"? I know it is recommended by ES developers to use == false at the place ! and using == true, but the check is nagging about it. I ignore it?

Yes please, Use `== true` and `== false` as this is the pattern in our codebase.

Please also merge master once again if you can (no rebase - force push).

**Comment by codebird** [comment]
Thanks for all the help @matriv hopefully it passes :)

**Comment by codebird** [comment]
hmmmm the only remaining issue with the checkStyle is the `== true`

**Comment by matriv** [comment]
Yep, my bad, sorry, please remove the `== true`. Only the `== false` is used instead of `!<expression>`.

**Review by matriv — APPROVED** [review]
LGTM

**Review by astefan — APPROVED** [review]
This LGTM. Thank you for doing all of this and being so patient.
I hate to be the "party" breaker here, but I would like to see one more test, that @matriv also suggested in one of his reviews, but from what I can see, it's missing: https://github.com/elastic/elasticsearch/pull/41964#discussion_r376820590

I tested the following query and it works, but I would also like to have it in the tests: `SELECT 144 / 12 AS division, 1, 2 AS x, 1 + 2 FROM test_emp GROUP BY gender, salary`.

Thanks again.

**Comment by codebird** [comment]
@astefan the test fails because of the column name of 1 + 2 H2 names it 3, ES names it 1 + 2 that's why I added an alias to the test. Shall we open another issue concerning this, or this is acceptable?

**Comment by astefan** [comment]
@codebird an alias is fine, thank you. No need for another issue, as I believe we are doing the right thing by keeping the column name as `1 + 2`.

**Comment by matriv** [comment]
@elasticmachine run elasticsearch-ci/2

**Comment by matriv** [comment]
@codebird I'd like to ask you for one more change please.
After some further investigation we came up with a bit different approach on the fix regarding
the resolution of attributes. The logic is the same though as in your fix: try to resolve groupings and/or aggregates based on the child plan (for example when there is a WHERE clause and the child plan is a filter). So instead of this code: https://github.com/elastic/elasticsearch/pull/41964/files#diff-bb55908282831d2f432f4a4650d55521R345
could  you please add the following:
```
@@ -733,6 +720,19 @@ public class Analyzer extends RuleExecutor<LogicalPlan> {
                 }
             }

+            // Try to resolve aggregates and groupings based on the child plan
+            if (plan instanceof Aggregate) {
+                Aggregate a = (Aggregate) plan;
+                LogicalPlan child = a.child();
+                List<Expression> newGroupings = new ArrayList<>(a.groupings().size());
+                a.groupings().forEach(e -> newGroupings.add(tryResolveExpression(e, child)));
+                List<NamedExpression> newAggregates = new ArrayList<>(a.aggregates().size());
+                a.aggregates().forEach(e -> newAggregates.add(tryResolveExpression(e, child)));
+                if (newAggregates.equals(a.aggregates()) == false || newGroupings.equals(a.groupings()) == false) {
+                    return new Aggregate(a.source(), child, newGroupings, newAggregates);
+                }
+            }
+
             return plan;
         }
```

**Comment by codebird** [comment]
much more elegant indeed.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java [hunk 19]
```diff
@@ -1130,7 +1130,8 @@ protected LogicalPlan rule(LogicalPlan plan) {
 
             plan.forEachDown(a -> {
```
Comment (codebird) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: @matriv I don't think this whole thing is still needed from 1131 till line 1141.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 20]
```diff
@@ -632,8 +620,7 @@ protected LogicalPlan rule(LogicalPlan plan) {
                     maybeResolved.add(or.resolved() ? or : tryResolveExpression(or, child));
                 }
 
-                Stream<Order> referencesStream = maybeResolved.stream()
-                        .filter(Expression::resolved);
+                Stream<Order> referencesStream = maybeResolved.stream().filter(Expression::resolved);
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: Please revert formatting changes.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 21]
```diff
@@ -706,8 +693,7 @@ protected LogicalPlan rule(LogicalPlan plan) {
                 Expression maybeResolved = tryResolveExpression(f.condition(), f.child());
 
                 AttributeSet resolvedRefs = new AttributeSet(maybeResolved.references().stream()
-                        .filter(Expression::resolved)
-                        .collect(toList()));
+                    .filter(Expression::resolved).collect(toList()));
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: same here

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java [hunk 22]
```diff
@@ -719,8 +705,9 @@ protected LogicalPlan rule(LogicalPlan plan) {
                     // resolution failed and the failed expressions might contain resolution information so copy it over
                     if (!failedAttrs.isEmpty()) {
                         // transform the orders with the failed information
-                        Expression transformed = f.condition().transformUp(ua -> resolveMetadataToMessage(ua, failedAttrs, "filter"),
-                                UnresolvedAttribute.class);
+                        Expression
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Analyzer.java]: and here

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java (hunk 19 above)
Comment (codebird) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: @matriv I don't think this whole thing is still needed from 1131 till line 1141.

Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: This should kick in for queries that end up to an LocalExec query, so no actual ES query is executed but one row (with literals only) is returned.

Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: But probably there is no such case as for group by.
If you remove it nothing breaks?

Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: nope, all integration tests are fine, I still need to run the unit tests make sure all is ok. will let you know.

Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Thank you!

Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: some unit tests fail. Example this: `testFoldingToLocalExecWithAggregationAndOrderByAndLimit`, only this `assertEquals(LocalExec.class, p.getClass());` as we removed it, but the result is good. I guess let's just keep it in

Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: Yes please. Could you check and revert those formatting changes I mentioned?

Comment (codebird) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/optimizer/Optimizer.java]: yep sorry, that was intellij.

**Comment by matriv** [comment]
Thanks a lot once again @codebird for your patience and your overall effort on this contribution!

**Comment by codebird** [comment]
Np.thanks for the good mentoring @matriv
